src/Service.GraphQLBuilder/GraphQLUtils.cs

```diff
@@ -270,5 +272,63 @@ public static ObjectType UnderlyingGraphQLEntityType(IType type)
                         innerException: error);
             }
         }
+
+        /// <summary>
+        /// Generates the entity name from the GraphQL context.
+        /// </summary>
+        /// <param name="context">Middleware context.</param>
+        /// <returns></returns>
+        public static string GetDataSourceNameFromGraphQLContext(IMiddlewareContext context, RuntimeConfig runtimeConfig)
+        {
+            string rootNode = context.Selection.Field.Coordinate.TypeName.Value;
+            string dataSourceName;
+
+            if (string.Equals(rootNode, "mutation", StringComparison.OrdinalIgnoreCase) || string.Equals(rootNode, "query", StringComparison.OrdinalIgnoreCase))
+            {
+                // we are at the root query node - need to determine return type and store on context.
+                // Output type below would be the graphql object return type - Books,BooksConnectionObject.
+                IOutputType outputType = context.Selection.Field.Type;
+                string entityName = outputType.TypeName();
+
+                // Below is only needed if say we have an Array return type or items object for plural case.
+                ObjectType underlyingFieldType = GraphQLUtils.UnderlyingGraphQLEntityType(outputType);
+
+                // Example: CustomersConnectionObject - for get all scenarios.
+                if (QueryBuilder.IsPaginationType(underlyingFieldType))
+                {
+                    IObjectField subField = GraphQLUtils.UnderlyingGraphQLEntityType(context.Selection.Field.Type).Fields[QueryBuilder.PAGINATION_FIELD_NAME];
+                    outputType = subField.Type;
+                    underlyingFieldType = GraphQLUtils.UnderlyingGraphQLEntityType(outputType);
+                    entityName = underlyingFieldType.Name;
+                }
+
+                // if name on schema is different from name in config.
+                // Due to possibility of rename functionality, entityName on runtimeConfig could be different from exposed schema name.
+                if (GraphQLUtils.TryExtractGraphQLFieldModelName(underlyingFieldType.Directives, out string? modelName))
+                {
+                    entityName = modelName;
+                }
+
+                dataSourceName = runtimeConfig.GetDataSourceNameFromEntityName(entityName);
+
+                // Store dataSourceName on context for later use.
+                context.ContextData.TryAdd(GenerateDataSourceNameKeyFromPath(context), dataSourceName);
+            }
+            else
+            {
+                // Derive node from path - e.g. /books/{id} - node would be books.
+                // for this queryNode path we have stored the datasourceName needed to retrieve query and mutation engine of inner objects
+                object? obj = context.ContextData[GenerateDataSourceNameKeyFromPath(context)];
```

Comment:
Where is a key/value pair added for the key contents generated by `GenerateDataSourceNameKeyFromPath(context)`? if you are generating the key name here, how would it already be present in the ContextData dictionary?
your comment mentions "for this queryNode path we have stored the datasourceName": where does that occur?

Reply:
That occurs on in the first part of the if statement at the root level.